Service/AbstractIndexBuilder.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace Algolia\AlgoliaSearch\Service;
+
+use Algolia\AlgoliaSearch\Exceptions\AlgoliaException;
+use Algolia\AlgoliaSearch\Helper\AlgoliaHelper;
+use Algolia\AlgoliaSearch\Helper\ConfigHelper;
+use Algolia\AlgoliaSearch\Logger\DiagnosticsLogger;
+use Magento\Framework\App\Area;
+use Magento\Framework\App\Config\ScopeCodeResolver;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Store\Model\App\Emulation;
+
+abstract class AbstractIndexBuilder
```

Comment:
I like this introduction of an abstract class. I also think we should have it implement an `IndexBuilder` interface under the `Api` namespace that contains common build methods. The methods would also be abstract in this class.

Reply:
I thought about this but as we want to keep full compatibility with the old methods (for now) and those methods all have a different signature, I don't know how to regroup them under the same interface (but maybe there's a trick that I'm not aware of), I saw that people sometimes use an `$args` array for that but we come back to the fact that it's not compatible with the old methods)

Reply:
More on JIRA to elaborate with this whole abstraction thing (coupled with the other comments)

Reply:
Got it! Follow up thoughts in Jira as well.

Reply:
Slated for follow up